Allow preboot callbacks to be passed directly

Lobstrosity opened the pull request:
For discussion.

I’ve been looking into server-side rendering a `react` app built on `slay`. Step 1 in all the `react` SSR literature is to build a `server.js` bundle so backend has access to your transpiled components. If you try to create such a bundle with a `slay` app, `webpack` goes no further than your index file since `slay` `require`s your preboots via expressions.

This PR removes expression-`require`s where necessary and allows you to pass your preboots in explicitly like so:

```
new slay.App(path.join(__dirname), {
  preboots: require('./lib/preboots'),
  middlewares: require('./lib/middlewares'),
  routes: require('./lib/routes')
});
```

…which gives `webpack` the hooks it needs to bundle the app.

(The new functionality is completely optional. The ability to pass paths in for those properties is intact. Or you can still pass nothing to allow default behavior.)

jcrugzz commented:
@Lobstrosity I kind of like this actually but would like to get @indexzero's opinion

SivanMehta commented on .gitignore:
```diff
@@ -30,3 +30,7 @@ node_modules
 build
+
+# Locks
+yarn.lock
```
`yarn` isn't a dependency of this module so this line can be removed.

indexzero replied:
One of those "why not" additions to the `.gitignore`. (e.g. not everyone uses Mac OSX and we have `.DS_Store` above).

SivanMehta commented on .gitignore:
```diff
@@ -30,3 +30,7 @@ node_modules
+
+# Locks
+yarn.lock
+package-lock.json
```
More of a question than a comment, but what is the outlook of upgrading `slay` to `npm@5`? In that case `package-lock.json` would need to be included.

indexzero replied:
Would like to see this removed in a PR that adds a `.gitattributes` file to treat `package-lock.json` as a binary file.

SivanMehta commented on lib/preboot/require.js:
```diff
@@ -1,26 +0,0 @@
-    }
-    bootable(app, options, callback);
-  };
-};
```
Your PR description says that it:

> removes expression-requires where necessary and allows you to pass your preboots in explicitly

Are you intending to *remove* the` require` functionality or simply *augment* it? It may be helpful write some tests to ensure that this PR is not introducing a breaking change.

Lobstrosity replied:
Default behavior will still `require` the same files as before. However, new behavior allows you to pass preboot callbacks in directly so that `slay` doesn’t have to require them via expression.

Default behavior is already tested by the `Running application tests` block.

There was existing behavior to allow paths to be passed in. (That wasn’t tested before but is now.) New behavior is also tested. Those new tests are in the `Preboot sources` block.

Lobstrosity replied:
I had originally removed `builtins.require` but that’s not necessary. Added back in 65391da.

indexzero commented on lib/app.js:
```diff
@@ -80,9 +80,9 @@ App.bootstrap = function (app, root) {
-    routes: app.given.routes || path.join(root, 'lib', 'routes')
+    preboots: typeof app.given.preboots === 'string' ? app.given.preboots : path.join(root, 'lib', 'preboots'),
+    middlewares: typeof app.given.middlewares === 'string' ? app.given.middlewares : path.join(root, 'lib', 'middlewares'),
+    routes: typeof app.given.routes === 'string' ? app.given.routes : path.join(root, 'lib', 'routes')
```
Semantically is this the behavior we really want? In the case any of these are functions falling back to the default value is setting a default that doesn't represent the actual location of the logic. 

In other words: what's the use of supplying a default value if it is `ENOENT`?

Lobstrosity replied:
Would it be acceptable to leave any that are provided as functions `undefined` in `paths`?

indexzero replied:
That appears to be the correct approach from my perspective. At least this way we are not lying to the user w.r.t. files that exist on disk. e.g.:

``` js
app.paths = ['preboots', 'middlewares', 'routes'].reduce((acc, name) => {
  if (typeof app.given[name] === 'string') {
    acc[name] = app.given[name];
  }

  return acc; 
}, {});
```

Lobstrosity commented:
Updated:
* In the `paths` reducer, `typeof given !== 'function'` is used to mimic original behavior as closely as possible. (Would you like me to add validation to `throw` if any of the values are non-string/function?)
* In `app.before('setup')`, added `resolve` function to simplify the check/fallback.

indexzero commented on lib/app.js:
```diff
@@ -99,14 +103,20 @@ App.bootstrap = function (app, root) {
 
   app.before('setup', function (instance, options, done) {
-    builtins.require(app.paths.preboots)(app, options, function (err) {
+    var resolve = function (name) {
```
Nitpick:

``` js
function resolve(name) {
```
